modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java

```diff
@@ -322,6 +328,8 @@ private void onMessage(String nodeName, QueryStartRequest msg) {
         assert nodeName != null && msg != null;
 
         DistributedQueryManager queryManager = queryManagerMap.computeIfAbsent(msg.queryId(), key -> {
+            sqlSchemaManager.waitActualSchema(msg.lastVersion());
```

Comment:
I'm afraid this is not very god idea to block inside update closure, because you are basically blocking even those queries which don't require latest schema. Moreover, it's possible to avoid thread blocking at all: we need to schedule the task after required schema will be available

Reply:
If i correctly understand you, such approach is not working for now, or i miss you idea ? 
`        CompletableFuture<SchemaPlus> fut = sqlSchemaManager.waitActualSchema(msg.lastVersion());

        fut.thenApply((sch) -> {
            DistributedQueryManager queryManager = queryManagerMap.computeIfAbsent(msg.queryId(), key -> {
                BaseQueryContext ctx = createQueryContext(key, msg.schema(), msg.parameters());

                return new DistributedQueryManager(ctx);
            });

            return queryManager;
        }).whenComplete((mgr, e) -> mgr.submitFragment(nodeName, msg.root(), msg.fragmentDescription(), msg.txAttributes()));`

Reply:
done